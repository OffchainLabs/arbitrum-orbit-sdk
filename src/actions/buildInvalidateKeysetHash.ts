import { Chain, Hex, PrepareTransactionRequestParameters, PublicClient, Transport } from 'viem';
import { sequencerInboxABI } from '../contracts/SequencerInbox';
import {
  ActionParameters,
  PrepareTransactionRequestReturnTypeWithChainId,
  WithAccount,
  WithUpgradeExecutor,
} from '../types/Actions';
import { Prettify } from '../types/utils';
import { validateParentChainPublicClient } from '../types/ParentChain';
import { prepareUpgradeExecutorCallParameters } from '../prepareUpgradeExecutorCallParameters';

export type BuildInvalidateKeysetHashParameters<Curried extends boolean = false> = Prettify<
  WithUpgradeExecutor<
    WithAccount<
      ActionParameters<
        {
          keysetHash: Hex;
        },
        'sequencerInbox',
        Curried
      >
    >
  >
>;

export type BuildInvalidateKeysetHashReturnType = PrepareTransactionRequestReturnTypeWithChainId;

export async function buildInvalidateKeysetHash<TChain extends Chain | undefined>(
  client: PublicClient<Transport, TChain>,
  params: BuildInvalidateKeysetHashParameters,
): Promise<BuildInvalidateKeysetHashReturnType> {
  const validatedPublicClient = validateParentChainPublicClient(client);
  const { account, upgradeExecutor, sequencerInbox: sequencerInboxAddress, ...args } = params;

  const request = await client.prepareTransactionRequest({
    chain: client.chain,
    account,
    ...prepareUpgradeExecutorCallParameters({
      to: sequencerInboxAddress,
      upgradeExecutor,
      args: [args.keysetHash],
      abi: sequencerInboxABI,
      functionName: 'invalidateKeysetHash',
    }),
  } satisfies PrepareTransactionRequestParameters);

  return { ...request, chainId: validatedPublicClient.chain.id };
}
