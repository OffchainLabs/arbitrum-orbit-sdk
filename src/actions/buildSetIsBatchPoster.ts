import { Address, Chain, PrepareTransactionRequestParameters, PublicClient, Transport } from 'viem';
import { sequencerInboxABI } from '../contracts/SequencerInbox';
import {
  ActionParameters,
  PrepareTransactionRequestReturnTypeWithChainId,
  WithAccount,
  WithUpgradeExecutor,
} from '../types/Actions';
import { Prettify } from '../types/utils';
import { prepareUpgradeExecutorCallParameters } from '../prepareUpgradeExecutorCallParameters';
import { validateParentChainPublicClient } from '../types/ParentChain';

type Args = {
  batchPoster: Address;
};

export type BuildSetIsBatchPosterParameters<Curried extends boolean = false> = Prettify<
  WithUpgradeExecutor<WithAccount<ActionParameters<Args, 'sequencerInbox', Curried>>>
>;

export type BuildSetIsBatchPosterReturnType = PrepareTransactionRequestReturnTypeWithChainId;

export async function buildSetIsBatchPoster<TChain extends Chain | undefined>(
  client: PublicClient<Transport, TChain>,
  params: BuildSetIsBatchPosterParameters & { enable: boolean },
): Promise<BuildSetIsBatchPosterReturnType> {
  const validatedPublicClient = validateParentChainPublicClient(client);
  const { account, upgradeExecutor, sequencerInbox: sequencerInboxAddress, ...args } = params;

  const request = await client.prepareTransactionRequest({
    chain: client.chain,
    account,
    ...prepareUpgradeExecutorCallParameters({
      to: sequencerInboxAddress,
      upgradeExecutor,
      args: [args.batchPoster, args.enable],
      abi: sequencerInboxABI,
      functionName: 'setIsBatchPoster',
    }),
  } satisfies PrepareTransactionRequestParameters);

  return { ...request, chainId: validatedPublicClient.chain.id };
}

export async function buildEnableBatchPoster<TChain extends Chain | undefined>(
  client: PublicClient<Transport, TChain>,
  args: BuildSetIsBatchPosterParameters,
): Promise<BuildSetIsBatchPosterReturnType> {
  return buildSetIsBatchPoster(client, {
    ...args,
    enable: true,
  });
}

export async function buildDisableBatchPoster<TChain extends Chain | undefined>(
  client: PublicClient<Transport, TChain>,
  args: BuildSetIsBatchPosterParameters,
): Promise<BuildSetIsBatchPosterReturnType> {
  return buildSetIsBatchPoster(client, {
    ...args,
    enable: false,
  });
}
