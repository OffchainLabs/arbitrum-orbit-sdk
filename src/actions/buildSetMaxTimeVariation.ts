import { Chain, PrepareTransactionRequestParameters, PublicClient, Transport } from 'viem';
import { sequencerInboxABI } from '../contracts/SequencerInbox';
import {
  ActionParameters,
  PrepareTransactionRequestReturnTypeWithChainId,
  WithAccount,
  WithUpgradeExecutor,
} from '../types/Actions';
import { Prettify } from '../types/utils';
import { prepareUpgradeExecutorCallParameters } from '../prepareUpgradeExecutorCallParameters';
import { validateParentChainPublicClient } from '../types/ParentChain';

type Args = {
  delayBlocks: bigint;
  futureBlocks: bigint;
  delaySeconds: bigint;
  futureSeconds: bigint;
};
export type BuildSetMaxTimeVariationParameters<Curried extends boolean = false> = Prettify<
  WithUpgradeExecutor<WithAccount<ActionParameters<Args, 'sequencerInbox', Curried>>>
>;

export type BuildSetMaxTimeVariationReturnType = PrepareTransactionRequestReturnTypeWithChainId;

export async function buildSetMaxTimeVariation<TChain extends Chain | undefined>(
  client: PublicClient<Transport, TChain>,
  {
    account,
    upgradeExecutor,
    sequencerInbox: sequencerInboxAddress,
    params,
  }: BuildSetMaxTimeVariationParameters,
): Promise<BuildSetMaxTimeVariationReturnType> {
  const validatedPublicClient = validateParentChainPublicClient(client);

  const request = await client.prepareTransactionRequest({
    chain: client.chain,
    account,
    ...prepareUpgradeExecutorCallParameters({
      to: sequencerInboxAddress,
      upgradeExecutor,
      args: [params],
      abi: sequencerInboxABI,
      functionName: 'setMaxTimeVariation',
    }),
  } satisfies PrepareTransactionRequestParameters);

  return { ...request, chainId: validatedPublicClient.chain.id };
}
