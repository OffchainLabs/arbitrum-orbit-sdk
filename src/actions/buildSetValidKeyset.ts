import { Chain, Hex, PrepareTransactionRequestParameters, PublicClient, Transport } from 'viem';
import { sequencerInboxABI } from '../contracts/SequencerInbox';
import {
  ActionParameters,
  PrepareTransactionRequestReturnTypeWithChainId,
  WithAccount,
  WithUpgradeExecutor,
} from '../types/Actions';
import { Prettify } from '../types/utils';
import { validateParentChainPublicClient } from '../types/ParentChain';
import { prepareUpgradeExecutorCallParameters } from '../prepareUpgradeExecutorCallParameters';

export type BuildSetValidKeysetParameters<Curried extends boolean = false> = Prettify<
  WithUpgradeExecutor<
    WithAccount<
      ActionParameters<
        {
          keyset: Hex;
        },
        'sequencerInbox',
        Curried
      >
    >
  >
>;

export type BuildSetValidKeysetReturnType = PrepareTransactionRequestReturnTypeWithChainId;

export async function buildSetValidKeyset<TChain extends Chain | undefined>(
  client: PublicClient<Transport, TChain>,
  params: BuildSetValidKeysetParameters,
): Promise<BuildSetValidKeysetReturnType> {
  const validatedPublicClient = validateParentChainPublicClient(client);
  const { account, upgradeExecutor, sequencerInbox: sequencerInboxAddress, ...args } = params;

  const request = await client.prepareTransactionRequest({
    chain: client.chain,
    account,
    ...prepareUpgradeExecutorCallParameters({
      to: sequencerInboxAddress,
      upgradeExecutor,
      args: [args.keyset],
      abi: sequencerInboxABI,
      functionName: 'setValidKeyset',
    }),
  } satisfies PrepareTransactionRequestParameters);

  return { ...request, chainId: validatedPublicClient.chain.id };
}
