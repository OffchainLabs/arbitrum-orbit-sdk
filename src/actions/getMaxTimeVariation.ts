import { Chain, PublicClient, Transport } from 'viem';
import { sequencerInboxABI } from '../contracts/SequencerInbox';
import { ActionParameters } from '../types/Actions';

export type GetMaxTimeVariationParameters<Curried extends boolean = false> = ActionParameters<
  {},
  'sequencerInbox',
  Curried
>;

export type GetMaxTimeVariationReturnType = {
  delayBlocks: bigint;
  futureBlocks: bigint;
  delaySeconds: bigint;
  futureSeconds: bigint;
};

export async function getMaxTimeVariation<TChain extends Chain | undefined>(
  client: PublicClient<Transport, TChain>,
  { sequencerInbox }: GetMaxTimeVariationParameters,
): Promise<GetMaxTimeVariationReturnType> {
  const [delayBlocks, futureBlocks, delaySeconds, futureSeconds] = await client.readContract({
    abi: sequencerInboxABI,
    functionName: 'maxTimeVariation',
    address: sequencerInbox,
  });
  return {
    delayBlocks,
    futureBlocks,
    delaySeconds,
    futureSeconds,
  };
}
