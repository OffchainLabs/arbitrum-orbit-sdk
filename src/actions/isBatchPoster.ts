import { Address, Chain, PublicClient, ReadContractReturnType, Transport } from 'viem';
import { sequencerInbox } from '../contracts';
import { ActionParameters } from '../types/Actions';

type Args = {
  batchPoster: Address;
};
type SequencerInboxABI = typeof sequencerInbox.abi;
export type IsBatchPosterParameters<Curried extends boolean = false> = ActionParameters<
  Args,
  'sequencerInbox',
  Curried
>;

export type IsBatchPosterReturnType = ReadContractReturnType<SequencerInboxABI, 'isBatchPoster'>;

export async function isBatchPoster<TChain extends Chain | undefined>(
  client: PublicClient<Transport, TChain>,
  args: IsBatchPosterParameters,
): Promise<IsBatchPosterReturnType> {
  return client.readContract({
    abi: sequencerInbox.abi,
    functionName: 'isBatchPoster',
    address: args.sequencerInbox,
    args: [args.batchPoster],
  });
}
