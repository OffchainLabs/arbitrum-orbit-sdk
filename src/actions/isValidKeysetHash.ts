import { Chain, Hex, PublicClient, ReadContractReturnType, Transport } from 'viem';
import { sequencerInbox } from '../contracts';
import { ActionParameters } from '../types/Actions';

type Args = {
  keysetHash: Hex;
};
type SequencerInboxABI = typeof sequencerInbox.abi;

export type IsValidKeysetHashParameters<Curried extends boolean = false> = ActionParameters<
  Args,
  'sequencerInbox',
  Curried
>;

export type IsValidKeysetHashReturnType = ReadContractReturnType<
  SequencerInboxABI,
  'isValidKeysetHash'
>;

export async function isValidKeysetHash<TChain extends Chain | undefined>(
  client: PublicClient<Transport, TChain>,
  args: IsValidKeysetHashParameters,
): Promise<IsValidKeysetHashReturnType> {
  return client.readContract({
    abi: sequencerInbox.abi,
    functionName: 'isValidKeysetHash',
    address: args.sequencerInbox,
    args: [args.keysetHash],
  });
}
