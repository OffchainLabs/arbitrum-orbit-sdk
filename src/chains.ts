import { defineChain } from 'viem';
import {
  mainnet,
  holesky,
  arbitrum as arbitrumOne,
  arbitrumNova,
  sepolia,
  arbitrumSepolia,
} from 'viem/chains';

const nitroTestnodeL1 = defineChain({
  id: 1_337,
  network: 'Nitro Testnode L1',
  name: 'nitro-testnode-l1',
  nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:8545'],
    },
    public: {
      http: ['http://127.0.0.1:8545'],
    },
  },
  testnet: true,
});

const nitroTestnodeL2 = defineChain({
  id: 412_346,
  network: 'Nitro Testnode L2',
  name: 'nitro-testnode-l2',
  nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:8547'],
    },
    public: {
      http: ['http://127.0.0.1:8547'],
    },
  },
  testnet: true,
});

// the L3 deployed as part of the nitro-testnode framework uses a custom fee token
const nitroTestnodeL3 = defineChain({
  id: 333333,
  network: 'Nitro Testnode L3',
  name: 'nitro-testnode-l3',
  nativeCurrency: { name: 'AppTestToken', symbol: 'APP', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:3347'],
    },
    public: {
      http: ['http://127.0.0.1:3347'],
    },
  },
  testnet: true,
});

export const chains = [
  // mainnet
  mainnet,
  arbitrumOne,
  arbitrumNova,
  // testnet
  sepolia,
  holesky,
  arbitrumSepolia,
  // local nitro-testnode
  nitroTestnodeL1,
  nitroTestnodeL2,
  nitroTestnodeL3,
] as const;

export {
  // mainnet
  mainnet,
  arbitrumOne,
  arbitrumNova,
  // testnet
  sepolia,
  holesky,
  arbitrumSepolia,
  // local nitro-testnode
  nitroTestnodeL1,
  nitroTestnodeL2,
  nitroTestnodeL3,
};
