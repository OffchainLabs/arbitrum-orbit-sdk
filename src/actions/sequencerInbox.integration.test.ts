import { describe, it, expect } from 'vitest';
import {
  createRollupHelper,
  getInformationFromTestnode,
  getNitroTestnodePrivateKeyAccounts,
} from '../testHelpers';
import { Hex, createPublicClient, http, zeroAddress } from 'viem';
import { nitroTestnodeL2 } from '../chains';
import { getMaxTimeVariation } from './getMaxTimeVariation';
import { isBatchPoster } from './isBatchPoster';
import { sequencerInboxABI } from '../contracts/SequencerInbox';
import { isValidKeysetHash } from './isValidKeysetHash';
import { buildSetValidKeyset } from './buildSetValidKeyset';
import { buildSetMaxTimeVariation } from './buildSetMaxTimeVariation';
import { buildDisableBatchPoster, buildEnableBatchPoster } from './buildSetIsBatchPoster';
import { generatePrivateKey, privateKeyToAccount } from 'viem/accounts';
import { buildInvalidateKeysetHash } from './buildInvalidateKeysetHash';

const { l3SequencerInbox, l3BatchPoster, l3UpgradeExecutor } = getInformationFromTestnode();
const { l3TokenBridgeDeployer, deployer, l3RollupOwner } = getNitroTestnodePrivateKeyAccounts();

const client = createPublicClient({
  chain: nitroTestnodeL2,
  transport: http(),
});

describe('max time variation management', () => {
  const defaultMaxTimeVariation = {
    delayBlocks: 5_760n,
    delaySeconds: 86_400n,
    futureBlocks: 12n,
    futureSeconds: 3_600n,
  };
  it('getMaxTimeVariation successfully fetches max time variation', async () => {
    const result = await getMaxTimeVariation(client, {
      sequencerInbox: l3SequencerInbox,
    });
    expect(result).toEqual(defaultMaxTimeVariation);
  });

  it('buildSetMaxTimeVariation successfully set max time varation', async () => {
    async function setMaxTimeVariation(changes: {
      delayBlocks: bigint;
      futureBlocks: bigint;
      delaySeconds: bigint;
      futureSeconds: bigint;
    }) {
      const transactionRequest = await buildSetMaxTimeVariation(client, {
        sequencerInbox: l3SequencerInbox,
        upgradeExecutor: l3UpgradeExecutor,
        ...changes,

        account: l3RollupOwner.address,
      });
      const txHash = await client.sendRawTransaction({
        serializedTransaction: await l3RollupOwner.signTransaction(transactionRequest),
      });

      await client.waitForTransactionReceipt({ hash: txHash });
    }

    const changes = {
      delayBlocks: 2_880n,
      futureBlocks: 6n,
      delaySeconds: 43_200n,
      futureSeconds: 1_800n,
    };
    await setMaxTimeVariation(changes);
    const newResult = await getMaxTimeVariation(client, {
      sequencerInbox: l3SequencerInbox,
    });
    expect(newResult).toEqual(changes);

    await setMaxTimeVariation(defaultMaxTimeVariation);
    const finalResult = await getMaxTimeVariation(client, {
      sequencerInbox: l3SequencerInbox,
    });
    expect(finalResult).toEqual(defaultMaxTimeVariation);
  });
});

describe('batch poster management', () => {
  it('isBatchPoster successfully fetches whether or an address is a batch poster', async () => {
    const isNotBatchPosterAddress = await isBatchPoster(client, {
      sequencerInbox: l3SequencerInbox,
      batchPoster: zeroAddress,
    });
    expect(isNotBatchPosterAddress).toBeFalsy();
    const isBatchPosterAddress = await isBatchPoster(client, {
      sequencerInbox: l3SequencerInbox,
      batchPoster: l3BatchPoster,
    });
    expect(isBatchPosterAddress).toBeTruthy();
  });

  it('successfully enable or disable an address as batch poster', async () => {
    const randomAddress = privateKeyToAccount(generatePrivateKey()).address;
    const isRandomAddressBatchPoster = await isBatchPoster(client, {
      sequencerInbox: l3SequencerInbox,
      batchPoster: randomAddress,
    });
    expect(isRandomAddressBatchPoster).toBeFalsy();

    const enableTransactionRequest = await buildEnableBatchPoster(client, {
      sequencerInbox: l3SequencerInbox,
      upgradeExecutor: l3UpgradeExecutor,
      batchPoster: randomAddress,
      account: l3RollupOwner.address,
    });
    const enableTxHash = await client.sendRawTransaction({
      serializedTransaction: await l3RollupOwner.signTransaction(enableTransactionRequest),
    });
    await client.waitForTransactionReceipt({ hash: enableTxHash });
    const isRandomAddressBatchPosterAfterEnabling = await isBatchPoster(client, {
      sequencerInbox: l3SequencerInbox,
      batchPoster: randomAddress,
    });
    expect(isRandomAddressBatchPosterAfterEnabling).toBeTruthy();

    const disableTransactionRequest = await buildDisableBatchPoster(client, {
      sequencerInbox: l3SequencerInbox,
      upgradeExecutor: l3UpgradeExecutor,
      batchPoster: randomAddress,
      account: l3RollupOwner.address,
    });
    const disableTxHash = await client.sendRawTransaction({
      serializedTransaction: await l3RollupOwner.signTransaction(disableTransactionRequest),
    });
    await client.waitForTransactionReceipt({ hash: disableTxHash });
    const isRandomAddressBatchPosterAfterDisabling = await isBatchPoster(client, {
      sequencerInbox: l3SequencerInbox,
      batchPoster: randomAddress,
    });
    expect(isRandomAddressBatchPosterAfterDisabling).toBeFalsy();
  });
});

describe('keyset management', () => {
  const keysetBytes =
    '0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000002560000000000000002000000000000000201216006dcb5e56764bb72e6a45e6deb301ca85d8c4315c1da2efa29927f2ac8fb25571ce31d2d603735fe03196f6d56bcbf9a1999a89a74d5369822c4445d676c15ed52e5008daa775dc9a839c99ff963a19946ac740579874dac4f639907ae1bc69f0c6694955b524d718ca445831c5375393773401f33725a79661379dddabd5fff28619dc070befd9ed73d699e5c236c1a163be58ba81002b6130709bc064af5d7ba947130b72056bf17263800f1a3ab2269c6a510ef8e7412fd56d1ef1b916a1306e3b1d9c82c099371bd9861582acaada3a16e9dfee5d0ebce61096598a82f112d0a935e8cab5c48d82e3104b0c7ba79157dad1a019a3e7f6ad077b8e6308b116fec0f58239622463c3631fa01e2b4272409215b8009422c16715dbede5909060121600835f995f2478f24892d050daa289f8b6b9c1b185bcd28532f88d610c2642a2dc6f3509740236d33c3e2d9136aab17f819c8c671293bba277717762e8d1c1f7bac9e17dd28d2939a959bb38e500f9c11c38cebbc426e2dea97c40175a655d17400ae6c75ff49e884c79469249e70953258854b64fa8445c585ad45dc6dc6975501c6af7cff7074202c687f8a7bf1a3ac192689755f232275b4c8421b1a5669e9b904c29a292cdf961b783a7c0b4ce736900de4d8c63c5f85a65cb44af34bef840acef84ab75f44c4c9137610b68107aff3bbdcc19119c7a927c115b7b9bfb27d85c500ee77d13ec5a97a3ae6bf51d3b70a5502e8416de7b5eb8e9feee376411ca35c8a7f3f597c7606578cf96a4715ce5a35cf48e39c0a1faa2dee22d74e681900000000000000000000';
  async function deployAnyTrustChainWithKeyset(keyset: Hex) {
    const batchPosters = [deployer.address];
    const validators = [deployer.address];

    const { createRollupInformation } = await createRollupHelper({
      deployer: l3TokenBridgeDeployer,
      batchPosters,
      validators,
      nativeToken: zeroAddress,
      client,
    });

    const { sequencerInbox, upgradeExecutor } = createRollupInformation.coreContracts;
    const transactionRequest = await buildSetValidKeyset(client, {
      sequencerInbox: sequencerInbox,
      keyset,
      account: l3TokenBridgeDeployer.address,
      upgradeExecutor,
    });
    const transactionHash = await client.sendRawTransaction({
      serializedTransaction: await l3TokenBridgeDeployer.signTransaction(transactionRequest),
    });
    await client.waitForTransactionReceipt({ hash: transactionHash });
    const logs = await client.getContractEvents({
      address: sequencerInbox,
      abi: sequencerInboxABI,
      eventName: 'SetValidKeyset',
    });

    const keysetHash = logs.find((log) => log.args.keysetBytes === keyset)?.args.keysetHash;

    return {
      keysetHash,
      sequencerInbox,
      upgradeExecutor,
    };
  }

  it('isValidKeysetHash successfully fetches whether a hash is a valid keyset hash', async () => {
    const invalidKeysetHash = await isValidKeysetHash(client, {
      sequencerInbox: l3SequencerInbox,
      keysetHash: '0x0000000000000000000000000000000000000000000000000000000000000000',
    });
    expect(invalidKeysetHash).toBeFalsy();

    const { sequencerInbox, keysetHash } = await deployAnyTrustChainWithKeyset(keysetBytes);

    const result = await isValidKeysetHash(client, {
      sequencerInbox,
      keysetHash: keysetHash!,
    });
    expect(result).toBeTruthy();
  });

  it('successfully invalidate a keyset hash', async () => {
    const { sequencerInbox, keysetHash, upgradeExecutor } = await deployAnyTrustChainWithKeyset(
      keysetBytes,
    );

    const result = await isValidKeysetHash(client, {
      sequencerInbox,
      keysetHash: keysetHash!,
    });
    expect(result).toBeTruthy();

    const transactionRequest = await buildInvalidateKeysetHash(client, {
      keysetHash: keysetHash!,
      sequencerInbox,
      account: l3TokenBridgeDeployer.address,
      upgradeExecutor,
    });
    const txHash = await client.sendRawTransaction({
      serializedTransaction: await l3TokenBridgeDeployer.signTransaction(transactionRequest),
    });
    await client.waitForTransactionReceipt({ hash: txHash });

    const resultAfterChange = await isValidKeysetHash(client, {
      sequencerInbox,
      keysetHash: keysetHash!,
    });
    expect(resultAfterChange).toBeFalsy();
  });
});
