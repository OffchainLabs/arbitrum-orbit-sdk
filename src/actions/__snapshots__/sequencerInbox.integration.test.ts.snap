// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`max time variation management > getMaxTimeVariation successfully fetches max time variation 1`] = `
{
  "delayBlocks": 5760n,
  "delaySeconds": 86400n,
  "futureBlocks": 12n,
  "futureSeconds": 3600n,
}
`;
